spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+Common examples include selecting the right plural form, and variants for different person genders.
+
+There should be a single level of selection (not nested like in ICU MessageFormat).
+It needs to support multiple selectors.
+
+In this syntax, a list of N selectors is followed by a list of pairs where
+the first element of each pair is a list of N value literals and
+the second element of each pair is a pattern.
+A `_` is a wildcard value that always matches.
+The last variant must have a list of all wildcard values.
+```
+[{$count :plural offset=1 grouping=always} {$gender}]
+[1 female] {{$name} added you to her circles.}
+[1 male] {{$name} added you to his circles.}
+[1 _] {{$name} added you to their circles.}
+[_ _] {{$name} added you and {#count} others to their circles.}
+```
+
+Lists are enclosed in square brackets, reminiscent of Python lists.
+The opening `[` also distinguishes the selection syntax from a simple pattern.
+
+TODO: Decide whether to enclose each value literal in
+the same pair of delimiters as literals in placeholder (for consistency),
+or whether to make that optional.
+(The `[]` value list syntax already indicates that value literals are enclosed.)
+Some literals may require it if they contain spaces.
+The `_` should probably never be enclosed in literal delimiters.
+
+Selector syntax follows placeholder syntax,
+except that a function must be specified.
+For the purpose of selection, there are three types of functions:
+1. Select-and-format functions combine the two functionalities,
+   and the selection is informed by the formatting.
+   For example, selectors for plural variants
+   (different selectors for cardinal-number vs. ordinal-number variants)
+   have to take into account how the number is formatted.
+2. Format-only functions can be used as selectors via
+   simple string matching of their output with the variant values.
+3. Select-only functions select among variant values, but they cannot be used in pattern placeholders.
+
+There is a simple format-only function that can be used for simple string matching.
+TODO: Decide on a name for this format-only function. Consider `:string`.
+
+Inside a selection-variant pattern,
+there is a special placeholder syntax for inserting the formatting result of a select-and-format function.
+This placeholder only specifies the selector’s argument name with a distinct prefix.
+It must not specify a function.
+In the example above, the `{#count}` value is the input $count minus the offset,
+like the `#` in an ICU PluralFormat, which is the input to the plural rules evaluation.
+This is not allowed for argument names used in select-only functions.
+- TODO: Bike-shedding on the prefix character, shown as `#` here.
+
+Inside selected patterns,
+the selector argument variables must not be used with the normal `$` placeholder syntax –
+for example, the patterns in the preceding example must not use `{$count}`.
```

Comment:
So I can't write:

```
[{$count :plural}]
[=0] {You have no items in your cart}
[one] {You have {$count :number style=spellout} item in your cart}
[_] {You have {$count : number style=spellout} items in your cart}
```

This seems hard for users to understand. They passed the argument by name. Why can't the format it? It isn't like the value has been consumed by whatever selector ate it previously.

Reply:
For plurals in particular, the formatting and selection are tied at the hip. If the spelled-out version of the number does not work grammatically like the `:plural` select-and-format function expected, then you get unhappy users.

Reply:
Ignore the style. The point I'm making is that your text says I cannot use the variable `$count` and a different formatter after having used it with the plural selector.

Reply:
Yes, for the stated reasons. Don't give users rope to hang themselves if we can avoid it :-)

Reply:
There is a problem with the following
```
[{$count :plural}]
[=0] {You have no items in your cart}
[one] {You have {$count :number style=spellout} item in your cart}
[_] {You have {$count : number style=spellout} items in your cart}
```
The plural categories are tied to the hip with the formatting. With the input number 1.01d, in some languages the category is 'one' if the format is an integer, but 'other' if the format has one (or more) decimals. So you can't actually correctly compute the plural category until you've formatted.

There are two ways to solve this:
1. tie the plural category to the formatted value, by having the formatting information up front, or
2. require the formatting information to be identical for every instance of the placeholder (eg it is an error if they are different)

It actually works pretty nicely to have the formatter return the plural category as an (optional) byproduct of formatting, because an intermediate step to producing the formatted number is typically the exact data necessary to compute the plural category. So the cleanest is to have a syntax that draws on that in some way. There are of course a few ways to do that. One is to use an assignment, and the other would be to have the formatting options in the selector, eg
```
[{$count :number style=spellout}]
[=0] {You have no items in your cart}
[one] {You have {$count} item in your cart}
[_] {You have {$count} items in your cart}
```